docs/manual/src/udl/decorators.md

```diff
@@ -0,0 +1,242 @@
+
+```kotlin
+class DemoLib(
+    val demoRustObject: DemoRustObject,
+    val decorator: MyDemoDecorator,
+) {
+    fun doExpensiveThing() = decorator.onBackgroundThread {
+        demoRustObject.doExpensiveThing()
+    }
+
+    fun doCrashyThing() = decorator.withErrorReporter {
+        demoRustObject.doCrashyThing()
+    }
+
+    fun doConsequentialThing() = decorator.thenNotifyListeners() {
+        demoRustObject.doConsequentialThing()
+    }
+}
+```
+
+This is much better, but it's still looking a bit cut and pasty. Enter decorator objects.
+
+## Decorators to the rescue
+
+> Decorator objects take their name from [Python's decorator methods][py-decorators].
+>
+> In Pythonic terms, Uniffi's decorator objects are a collection of decorator functions.
+>
+> Swift and Kotlin don't provide functionality to capture arbitrary `*args` and call a function with those same `*args`, so
+> at this time, decorator objects aren't as powerful as Python decorators. Nevertheless, they can be still quite useful.
+
+[py-decorators]: https://www.python.org/dev/peps/pep-0318/#on-the-name-decorator
+
+### UDL
+
+In the UDL we:
+
+* Specify that `DemoLib` should be a decorated version of `DemoRustObject` using the `[Decorated=`] annotation
+* Specify how methods should by decorated using the `[CallsWith=]` annotation
+* Don't specify the class that holds the decorator functions.  This is done on a per-app basis in `uniffi.toml`.
+
+```webidl
+namespace demo {}
+
+[Decorated=DemoLib]
+interface DemoRustObject {
+    [CallsWith=onBackgroundThread]
+    void do_expensive_thing();
+
+    [Throws=RustyError, CallsWith=withErrorReporter]
+    void do_crashy_thing();
+
+    [CallsWith=thenNotifyListeners]
+    void do_consequential_thing();
+}
+```
+
+### Write your decorator class
+
+```kotlin
+package my.uniffi.bindings.package
+
+class MyDemoDecorator(
+    val backgroundScope: CoroutineContext,
+    val errorReporter: (e: Exception) -> Unit,
+    val listener: () -> Void
+) {
+    fun <T> onBackgroundThread(rustCall: () -> T) {
+        backgroundScope.launch {
+            rustCall()
+        }
+    }
+
+    fun <T> withErrorReporter(rustCall: () -> T) =
+        try {
+            rustCall()
+        } catch (e: Exception) {
+            errorReporter(e)
+        }
+
+    fun thenNotifyListeners(rustCall: () -> T) {
+        try {
+            rustCall()
+        } finally {
+            listener()
+        }
+    }
+}
+```
+
+### Specify the decorator class in your uniffi.toml
+
+```toml
+
+[bindings.kotlin.decorators.DemoRustObject]
+class_name = MyDemoDecorator
+```
+
+### Use the decorator
+
+The above code will make UniFFI define a `DemoLib` class that uses
+`MyDemoDecorator` to decorate calls to `DemoRustObject`.  The `DemoLib`
+constructor will input a `DemoRustObject` and `MyDemoDecorator`.
+
+For example:
+
+```kotlin
+val demoLib = DemoLib(
+   DemoRustObject(),
+   MyDemoDecorator(backgroundScope, errorReporter, listener))
```

Comment:
Ah I see, so actually the caller has to know about *all three* of the classes involved here and use their explicit names, so it wouldn't make sense to try to give the underlying Rust object an auto-generated name. It also means that you can't fully understand the generated API just by looking at the `.udl` file, because the name `MyDemoDecorator` only appears in the `uniffi.toml`.

This seems cognitively very heavy to me. I don't have a good suggestion for how to get out from under it though, assuming that the user needs to be able to customize the construction of a `MyDemoDecorator`.

Reply:
I agree that it seems like there should be a less complicated way, but I can't see a good way to do that without giving up on the goal of having the decorators be runtime customizable.  You're making me wonder if that tradeoff is worth it though.

If we decide to stick with this approach, maybe updating the example, starting with your suggestion about "lib" vs "object" could help.

Reply:
> I can't see a good way to do that without giving up on the goal of having the decorators be runtime customizable.

Same :-/

I still like the "decorator" terminology better than other alternatives I've seen, but to be honest, my familiarity with Python decorators is probably what led me to assume some of these names would be hidden from the user. I don't have a concrete change to suggest, just noting it out loud.